R/chromote.R

```diff
@@ -411,31 +411,55 @@ Chromote <- R6Class(
     },
 
     #' @description Close the [`Browser`] object
-    close = function() {
-      # Must be alive to be active so we cache value before closing process
-      is_active <- self$is_active()
-
-      if (self$is_alive()) {
-        if (is_active) {
-          # send a message to the browser requesting that it close
-          self$Browser$close()
-        } else {
-          # terminate the process
-          private$browser$close()
-        }
+    #' @param wait If an integer, waits a number of seconds for the process to
+    #'   exit, killing the process if it takes longer than `wait` seconds to
+    #'   close. Use `wait = TRUE` to wait for 10 seconds, or `wait = FALSE` to
+    #'   close the connection without waiting for the process to exit. Only
+    #'   applies when Chromote is connected to a local process.
+    close = function(wait = TRUE) {
+      if (!isFALSE(wait)) {
+        if (isTRUE(wait)) wait <- 10
+        check_number_whole(wait, min = 0)
       }
 
-      if (is_active) {
-        private$ws$close()
+      is_local <- private$browser$is_local()
+
+      if (!is_local || !self$is_alive()) {
+        # For remote connections or cases where the process is already closed,
+        # we just close the websocket. Note that we skip $is_active() because it
+        # requires $is_alive().
+        if (private$ws$readyState() %in% c(0L, 1L)) {
+          private$ws$close()
+        }
+        return(invisible())
       }
 
-      invisible()
-    },
+      # close the browser nicely
+      self$Browser$close()
+
+      if (!isFALSE(wait)) {
+        # or close it forcefully if it takes too long
+        tryCatch(
+          {
+            pid <- private$browser$get_process()$get_pid()
+            message("waiting for browser process ", pid, " to shut down")
+            private$browser$get_process()$wait(timeout = wait)
+            message("browser process ", pid, " exited")
+            if (private$browser$get_process()$is_alive()) {
+              message("browser process ", pid, " is still alive >:(")
+              stop("shut it down")
+            }
```

Comment:
This is possible? I thought `$wait(timemout=)` would throw if time when over the timeout time

Reply:
I thought so too, but what's been happening in CI is that the `$wait()` exits without throwing, possibly before the timeout is reached, but the process is still alive. But maybe `$wait()` means something else in this context?